src/GraphQL/Server.hs

```diff
@@ -101,32 +110,39 @@ class ReadValue a where
 -- TODO not super hot on individual values having to be instances of
 -- HasGraph but not sure how else we can nest either types or
 -- (Object _ _ fields). Maybe instead of field we need a "SubObject"?
-instance forall m. (MonadThrow m, MonadIO m) => HasGraph m Int32 where
+instance forall m. (Monad m) => HasGraph m Int32 where
   type Handler m Int32 = m Int32
   -- TODO check that selectionset is empty (we expect a terminal node)
-  buildResolver handler _ =  map GValue.toValue handler
+  buildResolver handler _ =  do
+    map (ok . GValue.toValue) handler
 
-
-instance forall m. (MonadThrow m, MonadIO m) => HasGraph m Double where
+instance forall m. (Monad m) => HasGraph m Double where
   type Handler m Double = m Double
   -- TODO check that selectionset is empty (we expect a terminal node)
-  buildResolver handler _ =  map GValue.toValue handler
+  buildResolver handler _ =  map (ok . GValue.toValue) handler
 
-instance forall m. (MonadThrow m, MonadIO m) => HasGraph m Text where
+instance forall m. (Monad m) => HasGraph m Text where
   type Handler m Text = m Text
   -- TODO check that selectionset is empty (we expect a terminal node)
-  buildResolver handler _ =  map GValue.toValue handler
+  buildResolver handler _ =  map (ok . GValue.toValue) handler
 
 
-instance forall m hg. (MonadThrow m, MonadIO m, HasGraph m hg) => HasGraph m (List hg) where
+instance forall m hg. (Monad m, HasGraph m hg) => HasGraph m (List hg) where
   type Handler m (List hg) = [Handler m hg]
-  buildResolver handler selectionSet =
-    let a = traverse (flip (buildResolver @m @hg) selectionSet) handler
-    in map GValue.toValue a
-
-instance forall m ks enum. (MonadThrow m, MonadIO m, GraphQLEnum enum) => HasGraph m (Enum ks enum) where
+  buildResolver handler selectionSet = map resultValues a
+    where
+      a = traverse (flip (buildResolver @m @hg) selectionSet) handler
+      -- Aggregate results. TODO: Result should probably be a monoid
+      -- so we can collect values but I'm not sure what the monoid of
+      -- Value itself is.
+      resultValues :: [Result] -> Result
+      resultValues r =
+        let (errs, valueList) = foldl' (\(eb, vb) (Result ea va) -> ((eb <> ea), va:vb)) ([], []) r
+        in Result errs (GValue.toValue valueList)
```

Comment:
I think this should probably be a top-level function that lives next to the definition of `Result`.